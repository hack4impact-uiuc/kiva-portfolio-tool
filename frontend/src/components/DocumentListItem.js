import React, { Component } from 'react'
import { withRouter, Link } from 'react-router-dom'
import Dropzone from 'react-dropzone'
import { Row, Col, Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap'
import Iframe from 'react-iframe'

import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { updateDocuments, beginLoading, endLoading } from '../redux/modules/user'

import DocumentPreview from './DocumentPreview'

import {
  downloadDocument,
  uploadDocument,
  getDocumentsByUser,
  deleteDocument
} from '../utils/ApiWrapper'

import uploadImg from '../media/greyUpload.png'
import downloadImg from '../media/downloadGrey.png'
import visit from '../media/visit.png'
import remove from '../media/remove.png'

const mapStateToProps = state => ({
  isPM: state.user.isPM
})

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      beginLoading,
      endLoading,
      updateDocuments
    },
    dispatch
  )
}

/**
 * Shows each individual document in the Document List Component.
 * You can select this component to view, delete, or upload documents
 */
export class DocumentListItem extends Component {
  constructor(props) {
    super(props)

    this.state = {
      files: [],
      deleteModal: false,
      docClassModal: false
    }

    this.handleDownloadClick = this.handleDownloadClick.bind(this)
    this.onDrop = this.onDrop.bind(this)
    this.handleDelete = this.handleDelete.bind(this)
    this.deleteToggle = this.deleteToggle.bind(this)
    this.docClassToggle = this.docClassToggle.bind(this)
  }

  /**
   * Called upon uploading a file to a requirement
   */
  async onDrop(files) {
    // The MIME types of common filetypes
    const docTypes = [
      'application/msword',
      'application/vnd.ms-excel',
      'application/vnd.ms-powerpoint',
      'application/pdf',
      'image/jpeg',
      'image/png',
      'text/plain'
    ]

    if (!docTypes.includes(files[0].type)) {
      window.alert('Document type not acceptable')
    } else {
      this.setState({
        files
      })
      this.props.beginLoading()
      await uploadDocument(
        this.props.document.userID,
        this.state.files[0],
        this.state.files[0].name,
        this.props.document._id
      )
      const documents = await getDocumentsByUser(this.props.document.userID)
      if (documents) {
        this.props.updateDocuments(documents)
      } else {
        this.props.updateDocuments([])
      }
      this.props.endLoading()
    }
  }

  handleDownloadClick() {
    // Retrieve file from backend
    downloadDocument(document.fileID)
  }

  /**
   * Called when a user attempts to delete a document
   * The user will be prompted to confirm before the document is actually deleted
   */
  async handleDelete() {
    this.deleteToggle()
    this.props.beginLoading()
    await deleteDocument(this.props.document._id)
    const documents = await getDocumentsByUser(this.props.document.userID)
    if (documents) {
      this.props.updateDocuments(documents)
    } else {
      this.props.updateDocuments([])
    }
    this.props.endLoading()
  }

  deleteToggle() {
    this.setState({ deleteModal: !this.state.deleteModal })
  }

  docClassToggle() {
    this.setState({ docClassModal: !this.state.docClassModal })
  }

  render() {
    const { isPM } = this.props
    const customStyles = {
      height: '500px',
      width: '500px',
      overflow: 'scroll'
    }
    const externalCloseBtn = (
      <button
        className="close"
        style={{ position: 'absolute', top: '15px', right: '15px' }}
        onClick={this.toggle}
      >
        &times;
      </button>
    )

    return (
      <>
        <Modal isOpen={this.state.deleteModal} toggle={this.deleteToggle}>
          <ModalHeader>Are you sure you want to delete this document?</ModalHeader>
          <ModalFooter>
            <Button onClick={this.deleteToggle}>No</Button>
            <Button onClick={this.handleDelete}>Yes</Button>
          </ModalFooter>
        </Modal>
        <Modal
          isOpen={this.state.docClassModal}
          toggle={this.docClassToggle}
          external={externalCloseBtn}
        >
          <ModalHeader>{this.props.document.docClass.name}</ModalHeader>
          <ModalBody style={customStyles}>
            <p>{this.props.document.docClass.description}</p>
            <Iframe
              url={this.props.document.docClass.example}
              width="450px"
              height="500px"
              allowFullScreen
            />
          </ModalBody>
          <ModalFooter>
            <Button color="secondary" onClick={this.docClassToggle}>
              Close
            </Button>
          </ModalFooter>
        </Modal>
        <tr className="hoverable">
          <td data-testid="docClass">
            <Button className="add-doc-text" color="transparent" onClick={this.docClassToggle}>
              <span className="add-doc-text">{this.props.document.docClass.name}</span>
            </Button>
          </td>
          <td data-testid="fileName">
            {this.props.document.fileName ? this.props.document.fileName : 'N/A'}
          </td>
          <td className="interaction">
            {
              // little bit of a hack to coerce all images into one row inside one td element instead of having blank td elements taking up space
            }
            <Row className="button-tablerow">
              <Col sm="4" md="4">
                <DocumentPreview document={this.props.document} />
              </Col>
              <Col sm="4" md="4">
                {this.props.fileName ? (
                  <Button color="transparent">
                    <Link
                      to={{
                        pathname:
                          '/view/' + this.props.document.fileName + '/' + this.props.document._id,
                        state: { link: this.props.document.link }
                      }}
                    >
                      <img className="buttonimg" src={visit} alt="Visit icon" />
                    </Link>
                  </Button>
                ) : null}
              </Col>
              <Col sm="4" md="4">
                {isPM ? (
                  <Button color="transparent" onClick={this.deleteToggle}>
                    <img className="buttonimg" src={remove} width="23" alt="Remove icon" />
                  </Button>
                ) : (
                  <Dropzone onDrop={this.onDrop}>
                    {({ getRootProps, getInputProps }) => (
                      <section>
                        <div {...getRootProps()}>
                          <input {...getInputProps()} />
                          <img className="buttonimg" src={uploadImg} alt="Upload icon" />
                        </div>
                      </section>
                    )}
                  </Dropzone>
                )}
              </Col>
            </Row>
          </td>
        </tr>
      </>
    )
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(DocumentListItem))
