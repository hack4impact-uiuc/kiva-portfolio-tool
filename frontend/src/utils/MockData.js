const MockData = {
  missing: [
    {
      status: 'Missing',
      docClass: 'Board of Directors'
    },
    {
      status: 'Missing',
      docClass: 'Auditor Letter'
    }
  ],
  pending: [
    {
      status: 'Pending',
      docClass: 'Income Statement',
      fileName: 'income_statement.pdf',
      uploadDate: '12/4/2018'
    },
    {
      status: 'Pending',
      docClass: 'Balance Sheet',
      fileName: 'balance_sheet.pdf',
      uploadDate: '1/24/2019'
    }
  ],

  rejected: [
    {
      status: 'Rejected',
      docClass: 'Strategic Plan',
      fileName: 'strategic_plan.pdf',
      uploadDate: '12/24/2018'
    },
    {
      status: 'Rejected',
      docClass: 'Annual Plan',
      fileName: 'annual_plan.pdf',
      uploadDate: '1/3/2019'
    }
  ],
  approved: [
    {
      status: 'Approved',
      docClass: 'Financial Projections',
      fileName: 'financial_proj.pdf',
      uploadDate: '12/24/2018'
    },
    {
      status: 'Approved',
      docClass: 'Organizational Chart',
      fileName: 'org_chard.pdf',
      uploadDate: '1/23/2019'
    }
  ]
}

export default MockData
