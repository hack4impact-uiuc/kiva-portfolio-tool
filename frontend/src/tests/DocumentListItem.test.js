import React from 'react'
import DocumentListItem from '../components/DocumentListItem'
import Adapter from 'enzyme-adapter-react-16'
import { shallowToJson } from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import MockData from '../utils/MockData'

beforeAll(() => {
  configure({ adapter: new Adapter() })
})

test('DocumentListItem renders correctly', () => {
  const component = shallow(<DocumentListItem document={MockData.missing[0]} />)

  let tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})
