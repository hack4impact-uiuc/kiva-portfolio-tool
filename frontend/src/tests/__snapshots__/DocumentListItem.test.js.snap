// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentListItem renders correctly 1`] = `
<Fragment>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[Function]}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      wrapTag="div"
    >
      Are you sure you want to delete this document?
    </ModalHeader>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        No
      </Button>
      <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Yes
      </Button>
    </ModalFooter>
  </Modal>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    external={
      <button
        className="close"
        style={
          Object {
            "position": "absolute",
            "right": "15px",
            "top": "15px",
          }
        }
      >
        Ã—
      </button>
    }
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    role="dialog"
    toggle={[Function]}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      wrapTag="div"
    />
    <ModalBody
      style={
        Object {
          "height": "500px",
          "overflow": "scroll",
          "width": "500px",
        }
      }
      tag="div"
    >
      <p />
      <Iframe
        allowFullScreen={true}
        height="500px"
        width="450px"
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Close
      </Button>
    </ModalFooter>
  </Modal>
  <tr
    className="hoverable"
  >
    <td
      data-testid="docClass"
    >
      <Button
        className="add-doc-text"
        color="transparent"
        onClick={[Function]}
        tag="button"
      >
        <span
          className="add-doc-text"
        />
      </Button>
    </td>
    <td
      data-testid="fileName"
    >
      N/A
    </td>
    <td
      className="interaction"
    >
      <Connect(HOC)
        document={
          Object {
            "docClass": "Board of Directors",
            "status": "Missing",
          }
        }
      />
    </td>
    <td
      className="interaction"
      data-testid="interaction"
    />
    <td
      className="interaction padding-right-sm"
      data-testid="interaction"
    >
      <Dropzone
        onDrop={[Function]}
      >
        <Component />
      </Dropzone>
    </td>
  </tr>
</Fragment>
`;
